# Homework 11

In this assignment, we're continuing to work with the Sakila database that you installed for your previous Homework assignment. The database contains records of a ficticious chain of DVD video rental stores. Download the file "sqlite-sakila.db" from Canvas. If you put the database file into the same folder where you have saved this notebook, you can run the code provided below unchanged. Otherwise, you have to set a path to where you have saved the file. You can also find an ER-schema of the Sakila database on Canvas. Look at the ER-schema to understand the structure of the Sakila database. 

For all problems below, if you are not able to write `SQLite3` code to obtain results, try to write at least the corresponding SQL code statement in the provided Markdown box. If you can't run the code, you won't be able to get answers, of course, but you could at least get partial credit for your SQL statements. 

### Problem 1

Find the full names (first name and last name) of all customers whose first name begins with 'Ed'. Display them in a table. 

In [1]:
import sqlite3
import pandas as pd

def run_query(q):
    with sqlite3.connect("sqlite-sakila.db") as con:
            # Change the path to where YOU have saved the sqlite-sakila.db file
        return pd.read_sql(q, con)
    
### BEGIN SOLUTION
run_query("SELECT first_name, last_name FROM customer WHERE first_name LIKE 'ED%'")
### END SOLUTION 

first_name last_name
0       EDNA      WEST
1      EDITH  MCDONALD
2     EDWARD     BAUGH
3      EDWIN      BURK
4      EDDIE    TOMLIN
5      EDGAR    RHOADS
6    EDUARDO     HIATT

SQL code:

```
SELECT first_name, last_name 
    FROM customer
    WHERE first_name LIKE 'ED%';
```

Answer:

| first_name | last_name |
|:-- | :--|
| EDNA       | WEST      |
| EDITH      | MCDONALD  |
| EDWARD     | BAUGH     |
| EDWIN      | BURK      |
| EDDIE      | TOMLIN    |
| EDGAR      | RHOADS    |
| EDUARDO    | HIATT     |

### Problem 2

If the film titles are ordered alphabetically, which three movies come alphabetically first? 

In [2]:
### BEGIN SOLUTION 
run_query("SELECT title FROM film ORDER BY title LIMIT 3;")
### END SOLUTION 

title
0  ACADEMY DINOSAUR
1    ACE GOLDFINGER
2  ADAPTATION HOLES

SQL syntax:

```
SELECT title FROM film ORDER BY title LIMIT 3;
```

Answer:

Note, the films are already ordered alphabetically in the film table, which means that the `ORDER BY` statement could be omitted. 

| title                       |
|:--|
| ACADEMY DINOSAUR            |
| ACE GOLDFINGER              |
| ADAPTATION HOLES            |

### Problem 3

Produce a table of film titles, rental rates, ratings, and movie categories (by name, not by id number) for all G-rated Sci-Fi films. 

In [3]:
### BEGIN SOLUTION
run_query("SELECT f.title AS title, f.rental_rate AS rate, f.rating AS rating, c.name AS category \
    FROM film AS f\
    JOIN film_category AS fc \
    ON f.film_id = fc.film_id\
    JOIN category AS c\
    ON fc.category_id = c.category_id\
    WHERE c.name = 'Sci-Fi' AND f.rating = 'G'")
### END SOLUTION 

title  rate rating category
0        ANNIE IDENTITY  0.99      G   Sci-Fi
1       ARMAGEDDON LOST  0.99      G   Sci-Fi
2  BARBARELLA STREETCAR  2.99      G   Sci-Fi
3         CHEAPER CLYDE  0.99      G   Sci-Fi
4         CITIZEN SHREK  0.99      G   Sci-Fi
5   COLDBLOODED DARLING  4.99      G   Sci-Fi
6       DAISY MENAGERIE  4.99      G   Sci-Fi
7         GARDEN ISLAND  4.99      G   Sci-Fi
8            MARRIED GO  2.99      G   Sci-Fi
9      SUICIDES SILENCE  4.99      G   Sci-Fi

SQL syntax:

```
SELECT f.title AS title, f.rental_rate AS rate, f.rating AS rating, c.name AS category 
    FROM film AS f
    JOIN film_category AS fc 
    ON f.film_id = fc.film_id
    JOIN category AS c
    ON fc.category_id = c.category_id
    WHERE c.name = 'Sci-Fi' AND f.rating = 'G';
```

Answer:

| title                | rate | rating | category |
| :-- | :-- | :-- | :-- |
| ANNIE IDENTITY       | 0.99 | G      | Sci-Fi   |
| ARMAGEDDON LOST      | 0.99 | G      | Sci-Fi   |
| BARBARELLA STREETCAR | 2.99 | G      | Sci-Fi   |
| CHEAPER CLYDE        | 0.99 | G      | Sci-Fi   |
| CITIZEN SHREK        | 0.99 | G      | Sci-Fi   |
| COLDBLOODED DARLING  | 4.99 | G      | Sci-Fi   |
| DAISY MENAGERIE      | 4.99 | G      | Sci-Fi   |
| GARDEN ISLAND        | 4.99 | G      | Sci-Fi   |
| MARRIED GO           | 2.99 | G      | Sci-Fi   |
| SUICIDES SILENCE     | 4.99 | G      | Sci-Fi   |

### Problem 4

Some of the actors in the actor table have the same last names. Produce a table of last names of actors and counts of how many actors have that same last name (your table should have exactly the two columns "last_name" and "count").  Order the table first by counts in decreasing order and then alphabetically by last name and only display last names for cases where at least three actors share the same last name. 

In [4]:
### BEGIN SOLUTION 
run_query("SELECT last_name, COUNT(*) AS 'count'  \
    FROM actor \
    GROUP BY last_name \
    HAVING count(*) >=3 \
    ORDER BY count(*) DESC, last_name")
### END SOLUTION 

last_name  count
0      KILMER      5
1       NOLTE      4
2      TEMPLE      4
3      AKROYD      3
4       ALLEN      3
5       BERRY      3
6       DAVIS      3
7   DEGENERES      3
8     GARLAND      3
9     GUINESS      3
10     HARRIS      3
11    HOFFMAN      3
12    HOPKINS      3
13  JOHANSSON      3
14     KEITEL      3
15       PECK      3
16       TORN      3
17   WILLIAMS      3
18     WILLIS      3
19  ZELLWEGER      3

SQL syntax:

```
SELECT last_name, COUNT(*) AS 'count'  
    FROM actor 
    GROUP BY last_name 
    HAVING count(*) >=3 
    ORDER BY count(*) DESC, last name;
```

Answer:

| last_name | count |
|:-- | :--|
| KILMER    |     5 |
| NOLTE     |     4 |
| TEMPLE    |     4 |
| AKROYD    |     3 |
| ALLEN     |     3 |
| BERRY     |     3 |
| DAVIS     |     3 |
| DEGENERES |     3 |
| GARLAND   |     3 |
| GUINESS   |     3 |
| HARRIS    |     3 |
| HOFFMAN   |     3 |
| HOPKINS   |     3 |
| JOHANSSON |     3 |
| KEITEL    |     3 |
| PECK      |     3 |
| TORN      |     3 |
| WILLIAMS  |     3 |
| WILLIS    |     3 |
| ZELLWEGER |     3 |

### Problem 5

Write SQL code to return the count of the number of copies of the film 'WESTWARD SEABISCUIT' that are in the inventory. Your return table should have just one line (I don't want to have to pick out the answer out of a large table manually). 

In [5]:
### BEGIN SOLUTION 
run_query("SELECT f.title, COUNT(*) AS count\
    FROM film AS f\
    JOIN inventory as i\
    ON f.film_id = i.film_id\
    GROUP BY f.title\
    HAVING f.title = 'WESTWARD SEABISCUIT'")
### END SOLUTION 

title  count
0  WESTWARD SEABISCUIT      7

SQL syntax:

```
SELECT f.title, COUNT(*) AS count
    FROM film AS f
    JOIN inventory as i
    ON f.film_id = i.film_id
    GROUP BY f.title
    HAVING f.title = 'WESTWARD SEABISCUIT';
```  

Answer:

| title               | count |
|:-- |:-- |
| WESTWARD SEABISCUIT |     7 |